### PR TITLE
Port CRL test from boring crypto/x509/x509_test.cc; add fix to 1.1.0/master

### PR DESCRIPTION
**Pull request by richsalz**
To be ported to 1.1.0 and 1.0.2  Fails right now on master, needs to be investigated.

### CODE REVIEW

**Comment by FdaSilvaYY** [comment]
Error is pretty clear ;

test/crltest.c:40:5: error: string length '1253' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:64:5: error: string length '1298' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:77:5: error: string length '629' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:91:5: error: string length '658' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:105:5: error: string length '666' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:123:5: error: string length '654' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:140:5: error: string length '658' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

test/crltest.c:158:5: error: string length '682' is greater than the length '509' ISO C90 compilers are required to support [-Werror=overlength-strings]

**Comment by richsalz** [comment]
No, that's a build error which I did not see and which I will fix later.  The run-time error is this:
```
CRL with unknown critical extension (2) was accepted.
```

**Comment by davidben** [comment]
Please consider any work from BoringSSL included here to be covered by our existing CLA with OpenSSL.

I think the reason the test doesn't pass is because the fix didn't get applied to master (or 1.1.0), so the test is working. :-)

**Comment by davidben** [comment]
By the way, in case you need to generate more of these or whatever, here's the private key corresponding to `kCRLTestRoot`:
https://boringssl.googlesource.com/boringssl/+/f9f312af61f9ba87896736620d1e4e568c4442bd
(That probably should have been in a comment. I'll go fix that...)

Also, shameless plug, but der-ascii is really handy for generating these kinds of things:
https://github.com/google/der-ascii
You can see notes on how I fixed the signatures in samples/certificates.md (documented because I can never remember the incant for signing things)

**Comment by richsalz** [comment]
yes, right i forgot to push the commit with the fix.  (appropriate smiley)
the test gets picked to 1.0.2
everything here goes to 1.1.0 and master.
no longer a WIP.
ping @levitte who asked me about this off-line :)

**Comment by kroeckx** [comment]
You know that there are problems reported by travis, like "crltest: corrupted double-linked list"?

**Comment by richsalz** [comment]
No I did not notice the travis problems.

**Comment by richsalz** [comment]
Added a commit taht should fix this.  Thanks for pointing me to the problem.

File: test/crltest.c [hunk 1]
```diff
@@ -239,7 +240,11 @@ static int Verify(X509 *leaf, STACK_OF(X509) *roots, STACK_OF(X509_CRL) *crls,
     X509_STORE_CTX_set0_param(ctx, param);
 
     ERR_clear_error();
-    if (X509_verify_cert(ctx) != 1) {
+    retcode = X509_verify_cert(ctx);
+    /* Unset the CA and CRL stacks so they don't get free'd by us. */
+    X509_STORE_CTX_trusted_stack(ctx, NULL);
+    X509_STORE_CTX_set0_crls(ctx, NULL);
```
Comment (davidben) [inline test/crltest.c]: I'm not seeing where either of these fields are freed by `X509_STORE_CTX_free`, though perhaps I'm just not seeing it.

Anyway, this doesn't make sense. One would expect that either both `X509_STORE_CTX_set0_crls` and `X509_STORE_CTX_free` free the existing value or neither to. 

That is, if `ctx` owns `crls` as a result of line 235, then `ctx` is responsible for freeing it in both `set0_crls` and `free`. If `ctx` doesn't own `crls`, then `ctx` should not free it in either function. (At least from inspection, it seems to be the latter.)

Comment (richsalz) [inline test/crltest.c]: asan is complaining, so let's see what happens. :)

Comment (richsalz) [inline test/crltest.c]: doh, of course i wasn't making sense. :(  backed it out.  need to reproduce locally.

**Comment by richsalz** [comment]
found and fixed.  pushing new version.
unclear if i will backport this to 1.0.2

**Comment by snhenson** [comment]
You can avoid the PEM concatentation by using DER format and calling the relevant d2i function directly.

**Comment by richsalz** [comment]
ping @levitte to re-review.  uses the test framework.  for master only.

**Comment by richsalz** [comment]
ping @levitte please reconfirm since it uses the test harness now.  updated/rebased commit pushed.

File: test/crltest.c [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        goto err;
+
+    /* Create a stack; upref the cert because we free it below. */
+    X509_up_ref(root);
+    if (!sk_X509_push(roots, root))
+        goto err;
+
+    if (!X509_STORE_CTX_init(ctx, store, leaf, NULL))
+        goto err;
+    X509_STORE_CTX_trusted_stack(ctx, roots);
```
Comment (levitte) [inline test/crltest.c]: Change to:

```
    X509_STORE_CTX_set0_trusted_stack(ctx, roots);
```

Comment (levitte) [inline test/crltest.c]: (it will fix the Travis build issue)

**Comment by richsalz** [comment]
updated commit pushed.

**Comment by richsalz** [comment]
2b40699 in master, 0baae1c in 1.1.0

**Comment by FdaSilvaYY** [comment]
This fix was not pushed into 1.0.2 branch, who still has this issue too.
Any reason ?

**Comment by richsalz** [comment]
It was; see 3ade92e785bb3777c92332f88e23f6ce906ee260

**Pull request closed without merging**